domain/cloudimagemetadata/state/helpers_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state_test
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/cloudimagemetadata"
+)
+
+// retrieveMetadataFromDB retrieves all metadatas from the cloud_image_metadata database table.
+// It joins the architecture table to fetch architecture-related details and returns the metadata slice.
+//
+// It is used in test to keep save and find tests independent of each other
+func (s *stateSuite) retrieveMetadataFromDB() ([]cloudimagemetadata.Metadata, error) {
+	var metadatas []cloudimagemetadata.Metadata
+	return metadatas, s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.Query(`
+SELECT 
+created_at,
+source,
+stream,
+region,
+version,
+virt_type,
+root_storage_type,
+root_storage_size,
+priority,
+arch.name as archName,
+image_id
+ FROM cloud_image_metadata
+ JOIN architecture arch on cloud_image_metadata.architecture_id = arch.id`)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		defer func() { _ = rows.Close() }()
+		for rows.Next() {
+			var dbMetadata cloudimagemetadata.Metadata
+			if err = rows.Scan(
+				&dbMetadata.CreationTime,
+				&dbMetadata.Source,
+				&dbMetadata.Stream,
+				&dbMetadata.Region,
+				&dbMetadata.Version,
+				&dbMetadata.VirtType,
+				&dbMetadata.RootStorageType,
+				&dbMetadata.RootStorageSize,
+				&dbMetadata.Priority,
+				&dbMetadata.Arch,
+				&dbMetadata.ImageId,
+			); err != nil {
+				return errors.Trace(err)
+			}
+			metadatas = append(metadatas, dbMetadata)
+		}
```

Comment:
nit: We could use sqlair here.

Reply:
Yes, we could, but it would requires a loop to convert state.outputMetadata to cloudimagedata, so it won't really helps the readability amho.

And this way, it is completely independent from the rest of the tested code, which is better since this function is used to fetch data to assert.